server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -128,6 +131,11 @@ protected void setNextReader(LeafReaderContext ctx, int[] docsInLeaf) throws IOE
                 this.leafNestedDocuments = nestedDocuments.getLeafNestedDocuments(ctx);
                 this.leafStoredFieldLoader = storedFieldLoader.getLoader(ctx, docsInLeaf);
                 this.leafSourceLoader = sourceLoader.leaf(ctx.reader(), docsInLeaf);
+                if (idLoader != null) {
```

Comment:
I might have made `IdLoader` and interface with two implementations, just so that a reader of this code doesn't have to worry about it being nullable. It's just less work to read this part.